text/0000-machine-input.md

```diff
@@ -0,0 +1,183 @@
+
+greetService.send({ type: 'GREET.INPUT', name: 'David' });
+```
+
+Problems:
+- Since this input event is bespoke, there is no built-in way of identifying input properties
+- Input event _must_ be first event sent; this is not guaranteed
+- Cannot use input data at semantic start of machine; e.g., not usable in `entry`
+- No standard input event shape
+
+-----
+
+**Prior art:**
+
+- [Data in AWS Step Functions (States Language)](https://states-language.net/spec.html#data):
+
+> The interpreter passes data between states to perform calculations or to dynamically control the state machine’s flow. All such data MUST be expressed in JSON.
+>
+> When a state machine is started, the caller can provide an initial JSON text as input, which is passed to the machine's start state as input. If no input is provided, the default is an empty JSON object, `{}`. As each state is executed, it receives a JSON text as input and can produce arbitrary output, which MUST be a JSON text. [...]
+
+- [Azure Logic Apps](https://docs.microsoft.com/en-us/azure/logic-apps/logic-apps-overview) have _triggers_, which are events (with input data) that start a logic apps workflow.
+- [Google Cloud Platform Workflows](https://cloud.google.com/workflows/docs/passing-runtime-arguments) enable `params` that let you specify data in a workflow execution request
+- [Pipedream](https://pipedream.com/docs/workflows/steps/triggers/#app-based-triggers) has triggers that can contain input data for the start of a workflow.
+- [Argo workflow inputs](https://argoproj.github.io/argo-workflows/workflow-inputs/)
+- [GitHub Actions inputs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs)
+
+## Detailed design
+
+**Overall API**
+
+This RFC proposes to extend `{ type: 'xstate.init' }`, which is always the first event that a machine receives, with the optional `data?: ...` property:
+
+```ts
+const greetingMachine = createMachine({
+  schema: {
+    // For type-safety; can be used with JSON Schema
+    input: {} as { name: string }
+  },
+  entry: (_, event) => {
+    // event.type is 'xstate.init'
+    // event.data is { name: ... }
+    console.log(`Hello, ${event.data.name}!`);
+  }
+});
+
+// Non-interpreted environments
+const greetingMachineWithInput = greetingMachine
+  .withInput({ name: 'David' });
+
+greetingMachineWithInput.initialState.event.data;
+// { name: 'David' }
+
+// Interpreted environments
+interpret(greetingMachineWithInput).start();
+// logs 'Hello, David!'
+```
+
+### Technical Background
+
+The reason this was not implemented before is because SCXML does not define a separation between input data (params) and context data (datamodel). To provide "input data" to a statechart in SCXML, you provide [`<param>` values](https://www.w3.org/TR/scxml/#param) as [children inside of `<invoke>`](https://www.w3.org/TR/scxml/#N10FF5), which will overwrite the values in the `<datamodel>`.
```

Comment:
```suggestion
The reason this was not implemented before is because SCXML does not define a separation between input data (params) and context data (datamodel). To provide "input data" to a statechart in SCXML, you provide [`<param>` values](https://www.w3.org/TR/scxml/#param) as [children inside of `<invoke>`](https://www.w3.org/TR/scxml/#N10FF5), which will overwrite the values in the invokee's `<datamodel>`.
```